test/extension/handlers/topologymutation/handler.go

```diff
@@ -19,63 +19,225 @@ package topologymutation
+				return errors.Wrap(err, "error patching DockerClusterTemplate")
+			}
+		}
+		if controlPlaneTemplate, ok := obj.(*controlplanev1.KubeadmControlPlaneTemplate); ok {
+			if err := patchKubeadmControlPlaneTemplate(ctx, controlPlaneTemplate, variables); err != nil {
+				log.Error(err, "error patching KubeadmControlPlaneTemplateß")
+				return errors.Wrapf(err, "error patching KubeadmControlPlaneTemplate")
+			}
+		}
+		if machineDeploymentBootstrapTemplate, ok := obj.(*bootstrapv1.KubeadmConfigTemplate); ok {
+			if err := patchMachineDeploymentBootstrapTemplate(ctx, machineDeploymentBootstrapTemplate, variables); err != nil {
+				log.Error(err, "error patching KubeadmConfigTemplate")
+				return errors.Wrap(err, "error patching KubeadmConfigTemplate")
+			}
+		}
+		if dockerMachineTemplate, ok := obj.(*infrav1.DockerMachineTemplate); ok {
+			if err := patchDockerMachineTemplate(ctx, dockerMachineTemplate, variables); err != nil {
+				log.Error(err, "error patching DockerMachineTemplate")
+				return errors.Wrap(err, "error patching DockerMachineTemplate")
 			}
 		}
-
 		return nil
 	})
 }
 
-// patchDockerClusterTemplate patches the DockerClusterTepmlate.
-func patchDockerClusterTemplate(dockerClusterTemplate *infrav1.DockerClusterTemplate, templateVariables map[string]apiextensionsv1.JSON) error {
-	// Get the variable value as JSON string.
-	value, err := patchvariables.GetVariableValue(templateVariables, "lbImageRepository")
+func patchKubeadmControlPlaneTemplate(ctx context.Context, kubeadmControlPlaneTemplate *controlplanev1.KubeadmControlPlaneTemplate, templateVariables map[string]apiextensionsv1.JSON) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	// 1) Patch .Spec.KubeadmConfigSpec.ClusterConfiguration.ImageRepository with the value from the Cluster Topology variables.
+	//    If it's not set in the variables default to defaultImageRepository.
+	// TODO: no-op if the variable doesn't exist.
+	// TODO: error aggregation.
+	if found, imageRepository, err := getStringValueForVariable(templateVariables, "imageRepository"); found {
+		if err != nil {
+			return errors.Wrap(err, "could not set spec.template.spec.kubeadmConfigSpec.clusterConfiguration.imageRepository")
+		}
+		log.Info("Setting KubeadmControlPlaneTemplate imageRepository", "imageRepository", imageRepository)
+		kubeadmControlPlaneTemplate.Spec.Template.Spec.KubeadmConfigSpec.ClusterConfiguration.ImageRepository = imageRepository
+	}
+
+	// 2) If the Kubernetes version from builtin.controlPlane.version is below 1.24.0 set "cgroup-driver": "cgroupfs" to
+	//    - kubeadmConfigSpec.InitConfiguration.KubeletExtraArgs
+	//    - kubeadmConfigSpec.JoinConfiguration.KubeletExtraArgs
+	maximumVersionForPatch, err := version.ParseMajorMinorPatchTolerant("v1.24.0")
+	if err != nil {
+		return errors.Wrap(err, "could not set parse kubernetes version v1.24.0")
+	}
+	_, cpVersion, err := getStringValueForVariable(templateVariables, "builtin.controlPlane.version")
```

Comment:
We should probably handle the !found case as well

Reply:
I would really prefer if `getStringValueForVariable` could have the same behavior as the unstructured.NestedField* funcs.

Essentially:
* if not found => return false, "", nil
* if err => return false, "", err
* if found => return true, "value", nil

Feels more idiomatic to me, given that almost everywhere other return parameters are ignored as soon as an error was returned.

An alternative that is used in client-go is essential only returning value, err and then checking if err is a not found err (~ translated to our use case)